ddtrace/debugging/_function/discovery.py

```diff
@@ -0,0 +1,241 @@
+from types import ModuleType
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Type
+from typing import Union
+from typing import cast
+
+from ddtrace.internal.compat import PYTHON_VERSION_INFO as PY
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.module import origin
+from ddtrace.internal.safety import _isinstance
+from ddtrace.internal.utils.inspection import linenos
+
+
+log = get_logger(__name__)
+
+FunctionContainerType = Union[type, property, classmethod, staticmethod, Tuple, ModuleType]
+
+ContainerKey = Union[str, int, Type[staticmethod], Type[classmethod]]
+
+CONTAINER_TYPES = (type, property, classmethod, staticmethod)
+
+if PY < (3, 7):
+    # DEV: Prior to Python 3.7 the ``cell_content`` attribute of ``Cell``
+    # objects can only be mutated with the C API.
+    import ctypes
+
+    PyCell_Set = ctypes.pythonapi.PyCell_Set
+    PyCell_Set.argtypes = (ctypes.py_object, ctypes.py_object)
+    PyCell_Set.restype = ctypes.c_int
+
+    set_cell_contents = PyCell_Set
+else:
+
+    def set_cell_contents(cell, contents):  # type: ignore[misc]
+        cell.cell_contents = contents
+
+
+class FullyNamed(Protocol):
+    """A fully named object."""
+
+    __name__ = None  # type: Optional[str]
+    __fullname__ = None  # type: Optional[str]
+
+
+class FullyNamedFunction(FullyNamed):
+    """A fully named function object."""
+
+    def __call__(self, *args, **kwargs):
+        pass
+
+
+class ContainerIterator(Iterator, FullyNamedFunction):
+    """Wrapper around different types of function containers.
+
+    A container comes with an origin, i.e. a parent container and a position
+    within it in the form of a key.
+    """
+
+    def __init__(
+        self,
+        container,  # type: FunctionContainerType
+        origin=None,  # type: Optional[Union[Tuple[ContainerIterator, ContainerKey], Tuple[FullyNamedFunction, str]]]
+    ):
+        # type: (...) -> None
+        if isinstance(container, (type, ModuleType)):
+            self._iter = iter(container.__dict__.items())
+            self.__name__ = container.__name__
+
+        elif isinstance(container, tuple):
+            self._iter = iter(enumerate(_.cell_contents for _ in container))  # type: ignore[arg-type]
+            self.__name__ = "<locals>"
+
+        elif isinstance(container, property):
+            self._iter = iter(
+                (m, getattr(container, a)) for m, a in {("getter", "fget"), ("setter", "fset"), ("deleter", "fdel")}
+            )
+            assert container.fget is not None
+            self.__name__ = container.fget.__name__
+
+        elif isinstance(container, (classmethod, staticmethod)):
+            self._iter = iter([(type(container), container.__func__)])  # type: ignore[list-item]
+            self.__name__ = None
+
+        else:
+            raise TypeError("Unsupported container type: %s", type(container))
```

Comment:
Do we need to catch this somewhere and write a log or we'd break a customer's application? I'm not sure what standard practice is regarding raising errors and catching them from the tracer standpoint, but I could imagine a customer being annoyed that we're keeping their application from running.

Reply:
That's a good point. This error should be handled by the user of the `FunctionDiscovery` class. But, in reality, we should never see this exception being raised. I've put this here just for good measure, but ideally we'd never hit this branch.